src/boot/efi/boot.c

```diff
@@ -2334,6 +2469,55 @@ static VOID config_write_entries_to_variable(Config *config) {
         (void) efivar_set_raw(LOADER_GUID, L"LoaderEntries", buffer, sz, 0);
 }
 
+static EFI_STATUS secure_boot_discover_keys(Config *config, EFI_FILE *root_dir) {
+        EFI_STATUS err = EFI_SUCCESS;
+        _cleanup_(FileHandleClosep) EFI_FILE_HANDLE keys_basedir = NULL;
+
+        if (!setup_mode_enabled())
+                return EFI_SUCCESS;
+
+        /* the lack of a 'keys' directory is not fatal and is silently ignored */
+        err = uefi_call_wrapper(root_dir->Open, 5, root_dir, &keys_basedir, (CHAR16*) L"\\loader\\keys", EFI_FILE_MODE_READ, 0ULL);
+        if (err == EFI_NOT_FOUND)
+                return EFI_SUCCESS;
+        if (EFI_ERROR(err))
+                return err;
+
+        for (;;) {
+                CHAR16 buf[256];
+                UINTN bufsize = 256;
+                _cleanup_freepool_ CHAR16 *keys_dir = NULL;
+                EFI_FILE_INFO *f;
+                ConfigEntry *entry = NULL;
+
+                err = uefi_call_wrapper(root_dir->Read, 3, keys_basedir, &bufsize, buf);
+                if (bufsize == 0 || EFI_ERROR(err))
+                        break;
+
+                f = (EFI_FILE_INFO *) buf;
+                if (f->FileName[0] == '.')
+                        continue;
+
+                if (!(f->Attribute & EFI_FILE_DIRECTORY))
+                        continue;
+
+                entry = config_entry_add_signature_keys(root_dir, config, f->FileName);
+
+                /* if auto enrollement is activated (this is the default), we try
+                 * to load keys straight from the /loader/keys/auto location. If
+                 * it fails because the directory is not present we keep going
+                 * silently, otherwise we log an error and still proceed to add
+                 * the entries to the menu (including the auto one if it exists) */
+                if (config->auto_enroll == true && StrCmp(f->FileName, L"auto") == 0) {
+                        err = secure_boot_enroll(config, entry);
+                        if (!EFI_ERROR(err))
+                                return EFI_SUCCESS;
+                }
```

Comment:
instead of nesting the immediate enrolling here into the loop i think it would be nicer to maintain a local variable here "auto_keys" or so, and once we see the "auto" entry assign it the "entry" object we just allocated. And then once the loop is done, right before returning in this function check if "auto_keys" is set and if so, enroll it.

why? of course, the behaviour won't really change except we'll discover all keyrings before running the "auto" one, but we won't use them. So this is more conceptually nicer I think, simply because it separates our the pure enumeration from the execution a bit. i.e. we first enumerate, and once that is complete go for the execution. But merging these two bits in the same for loop sounds a bit too "dense" for me.